src/Symfony/Component/Config/Builder/ArrayShapeGenerator.php

```diff
@@ -0,0 +1,215 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Config\Builder;
+
+use Symfony\Component\Config\Definition\ArrayNode;
+use Symfony\Component\Config\Definition\BaseNode;
+use Symfony\Component\Config\Definition\BooleanNode;
+use Symfony\Component\Config\Definition\EnumNode;
+use Symfony\Component\Config\Definition\FloatNode;
+use Symfony\Component\Config\Definition\IntegerNode;
+use Symfony\Component\Config\Definition\NodeInterface;
+use Symfony\Component\Config\Definition\NumericNode;
+use Symfony\Component\Config\Definition\PrototypedArrayNode;
+use Symfony\Component\Config\Definition\ScalarNode;
+use Symfony\Component\Config\Definition\StringNode;
+use Symfony\Component\Config\Definition\VariableNode;
+
+/**
+ * @author Alexandre Daubois <alex.daubois@gmail.com>
+ *
+ * @internal
+ */
+final class ArrayShapeGenerator
+{
+    public const FORMAT_PHPDOC = 'phpdoc';
+    public const FORMAT_JETBRAINS_ATTRIBUTE = 'jetbrains_attribute';
+
+    /**
+     * @param self::FORMAT_* $format
+     */
+    public static function generate(ArrayNode $node, string $format): string
+    {
+        if (self::FORMAT_PHPDOC === $format) {
+            return self::prependPhpDocWithStar(self::doGeneratePhpDoc($node));
+        }
+
+        if (self::FORMAT_JETBRAINS_ATTRIBUTE === $format) {
+            return self::doGenerateJetBrainsArrayShape($node);
+        }
+
+        throw new \LogicException(\sprintf('Unsupported format to generate array shape. Expected one of "%s", got "%s".', implode('", "', [self::FORMAT_PHPDOC, self::FORMAT_JETBRAINS_ATTRIBUTE]), $format));
+    }
+
+    private static function doGeneratePhpDoc(NodeInterface $node, int $nestingLevel = 1): string
+    {
+        if (!$node instanceof ArrayNode) {
+            return $node->getName();
+        }
+
+        if ($node instanceof PrototypedArrayNode) {
+            $prototype = $node->getPrototype();
+            if ($prototype instanceof ArrayNode) {
+                return 'array<'.self::doGeneratePhpDoc($prototype, $nestingLevel).'>';
+            }
+
+            return 'array<'.self::handleNodeType($prototype).'>';
```

Comment:
the type of keys should also be documented here, depending on whether the node is expecting a map or no (i.e. whether `getKeyAttribute()` returns a non-null value)